Serverless Kubernetes deployments with AWS Fargate (#648)
ordina-jworks/ordina-jworks.github.io: _posts/2020-12-10-aws-fargate-serverless-deployments.md

fargate-ip-192-168-124-46.eu-west-1.compute.internal    Ready     <none>    5m       v1.18.8-eks-7c9bda
fargate-ip-192-168-97-139.eu-west-1.compute.internal    Ready     <none>    5m       v1.18.8-eks-7c9bda
```
This is because of the two CoreDNS pods that get spun up after the setup of the cluster.
```
$> kubectl get pods -n kube-system
NAME                     READY     STATUS    RESTARTS   AGE
coredns-58c89c64-pmjh4   1/1       Running   0          12m
coredns-58c89c64-rm4dr   1/1       Running   0          12m
```

Each pod gets its own Fargate node and represents the resources that the pods get in order to successfully function. 
When you create a new deployment, you will notice that you have a pod in the status pending creation.

```
$> kubectl create deployment spring-boot-docker --image springio/gs-spring-boot-docker
deployment.apps/spring-boot-docker created

$> kubectl get pods
NAME                                  READY     STATUS    RESTARTS   AGE
spring-boot-docker-6656b9d9fb-h82pk   0/1       Pending   0          5s
```

This is because Fargate has yet to create a virtual node for the Pod to run in.

> After a few second you can see that a new node has been added to your cluster.

Author: ```suggestion
After a few seconds you can see that a new node has been added to your cluster.
```

---
layout: post
authors: [yolan_vloeberghs]
title: "Serverless Kubernetes deployments with AWS Fargate"
image: /img/2020-12-10-aws-fargate-serverless-deployments.jpg
tags: [AWS, Fargate, Kubernetes]
category: Cloud
comments: true
---
# Prerequisites
 - An AWS account
 - kubectl, aws-cli & eksctl installed
 - Some Kubernetes knowledge

# Introduction
Introduced in late 2017, AWS Fargate is a compute engine for serverless container deployments. 
The engine allows you to run containers without worrying about your infrastructure.   
Node scaling and configuration is done by AWS, which means that you only have to worry about the health of your own resources, not of your Kubernetes infrastructure.  

Fargate eliminates the struggle of configuring, scaling and initializing nodes on its own.
Because as we know, managing clusters / servers / nodes can be a challenging and expensive task.

> Pricing is based on the CPU and RAM that your Pod allocates.

Reviewer: ```suggestion
Pricing is based on the CPU and RAM that is allocated to your pod.
```

> You don't pay for separate nodes anymore, but instead per Pod which could drastically decrease computing costs.

Reviewer: ```suggestion
Instead of paying for compute nodes you pay for resources allocated to your pods which could drastically decrease computing costs.
```

# Setup

> AWS Fargate introduced for Amazon ECS first in 2017 and then in 2019, they added compatibility for Amazon EKS.

Reviewer: ```suggestion
AWS Fargate was initially introduced for ECS in 2017. Later in 2019 they added support for EKS.
```

You can also create a new cluster with instant Fargate support. 
For simplicity, I will initialize a new cluster.
 
`eksctl create cluster --name <cluster-name> --region eu-west-1 --fargate`

The initialization process takes a while, so don't wait around for it and find something else to do in the meantime.
It shouldn't take more than 15 minutes.

# Fargate in action
After your cluster has been created, you should see two Fargate nodes already running.
```
$> kubectl get nodes 
NAME                                                    STATUS    ROLES     AGE       VERSION
fargate-ip-192-168-124-46.eu-west-1.compute.internal    Ready     <none>    5m       v1.18.8-eks-7c9bda
fargate-ip-192-168-97-139.eu-west-1.compute.internal    Ready     <none>    5m       v1.18.8-eks-7c9bda
```
This is because of the two CoreDNS pods that get spun up after the setup of the cluster.
```
$> kubectl get pods -n kube-system
NAME                     READY     STATUS    RESTARTS   AGE
coredns-58c89c64-pmjh4   1/1       Running   0          12m
coredns-58c89c64-rm4dr   1/1       Running   0          12m
```

> Each Pod gets its own Fargate node and represents the resources that the Pods get in order to successfully function. 

Reviewer: ```suggestion
Each pod gets its own Fargate node and represents the resources that the pods get in order to successfully function. 
```

> When you create a new Deployment, you will notice that you have a pending Pod creation.

Reviewer: ```suggestion
When you create a new deployment, you will notice that you have a pod in the status pending creation.
```

```
$> kubectl create deployment spring-boot-docker --image springio/gs-spring-boot-docker
deployment.apps/spring-boot-docker created

$> kubectl get pods
NAME                                  READY     STATUS    RESTARTS   AGE
spring-boot-docker-6656b9d9fb-h82pk   0/1       Pending   0          5s
```

This is because Fargate has yet to create a virtual node for the Pod to run in.

> Wait a few seconds.

Reviewer: ```suggestion

```

> Consulting the cluster's nodes will show that a new node has been added to your cluster.

Reviewer: ```suggestion
After a few second you can see that a new node has been added to your cluster.
```

```
$> kubectl get nodes 
NAME                                                    STATUS    ROLES     AGE       VERSION
fargate-ip-192-168-113-246.eu-west-1.compute.internal   Ready     <none>    16s       v1.18.8-eks-7c9bda
fargate-ip-192-168-124-46.eu-west-1.compute.internal    Ready     <none>    52m       v1.18.8-eks-7c9bda
fargate-ip-192-168-97-139.eu-west-1.compute.internal    Ready     <none>    52m       v1.18.8-eks-7c9bda
```

> Now your virtual node has been added, you will see your Pod starting up and fully functioning shortly.

Reviewer: ```suggestion
Now that your virtual node has been added you can see that your pod is starting up.
```

```
$> kubectl get pods
NAME                                  READY     STATUS    RESTARTS   AGE
spring-boot-docker-6656b9d9fb-h82pk   1/1       Running   0          2m26s
```

# Conclusion

> So right now we have a fully functioning Kubernetes cluster without having to touch or setup any node or configuration.

Reviewer 2: ```suggestion
So right now we have a fully functioning Kubernetes cluster without having to touch or set up any node or configuration.
```

> So right now we have a fully functioning Kubernetes cluster without having to touch or set up any node or configuration.

Reviewer: ```suggestion
So right now we have a fully functioning Kubernetes cluster without having to touch or set up any nodes or configuration.
```
